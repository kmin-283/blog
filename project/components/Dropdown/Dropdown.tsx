import React, { useEffect, useState, useRef, FC } from "react";
import { BsThreeDots } from "react-icons/bs";
import styles from "./Dropdown.module.css";

const Dropdown: FC = ({ children }) => {
  const [show, setShow] = useState<boolean>(false);
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const checkIfClickedOutside = (event: MouseEvent) => {
      if (show && ref.current && !ref.current?.contains(event.target as Node)) {
        setShow(!show);
      }
    };
    document.addEventListener("click", checkIfClickedOutside);
    return () => {
      document.removeEventListener("click", checkIfClickedOutside);
    };
  }, [show]);

  const showDropdown = () => {
    setShow(!show);
  };

  return (
    <section className={styles.dropdown} ref={ref}>
      <BsThreeDots
        className={styles.dropdownIcon}
        onClick={showDropdown}
        size={"2em"}
      />
      {show && <div className={styles.dropdownContent}>{children}</div>}
    </section>
  );
};
export default Dropdown;
