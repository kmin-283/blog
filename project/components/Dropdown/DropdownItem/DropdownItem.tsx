import React, { ReactChild, FC } from "react";
import styles from "./DropdownItem.module.css";

interface DropdownItemProps {
  onClick: () => void;
  icon: ReactChild;
}

const DropdownItem: FC<DropdownItemProps> = ({ children, onClick, icon }) => {
  return (
    <button className={styles.dropdownBtn} onClick={onClick}>
      <div className={styles.icon}>{icon}</div>
      <div className={styles.role}>{children}</div>
    </button>
  );
};

export default DropdownItem;
