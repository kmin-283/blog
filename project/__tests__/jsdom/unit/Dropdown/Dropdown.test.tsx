import React from "react";
import Dropdown from "@/components/Dropdown/Dropdown";
import DropdownItem from "@/components/Dropdown/DropdownItem/DropdownItem";
import {
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

describe("InpageNavigation 컴포넌트", () => {
  const playGround = document.createElement("div");
  const click1st = jest.fn(() => {
    playGround.innerText = "첫 번째 버튼이 클릭되었습니다";
  });
  const click2nd = jest.fn(() => {
    playGround.innerText = "두 번째 버튼이 클릭되었습니다";
  });
  let component: RenderResult;
  afterAll(cleanup);
  beforeEach(() => {
    component = render(
      <Dropdown>
        <DropdownItem onClick={click1st} icon={<span>dummyIcon</span>}>
          첫 번째 동작
        </DropdownItem>
        <DropdownItem onClick={click2nd} icon={<span>dummyIcon</span>}>
          두 번째 동작
        </DropdownItem>
      </Dropdown>
    );
  });

  test("렌더링", () => {
    const dropdown = renderer
      .create(
        <Dropdown>
          <DropdownItem onClick={click1st} icon={<span>dummyIcon</span>}>
            첫 번째 동작
          </DropdownItem>
          <DropdownItem onClick={click2nd} icon={<span>dummyIcon</span>}>
            두 번째 동작
          </DropdownItem>
        </Dropdown>
      )
      .toJSON();
    expect(dropdown).toMatchSnapshot();
  });

  test("dropdownIcon 클릭시 DropdownContent 생성", () => {
    const { container } = component;
    const dropdownIcon = container.querySelector(".dropdownIcon")!;

    expect(container.querySelector(".dropdownContent")).not.toBeInTheDocument();
    fireEvent.click(dropdownIcon);
    expect(container.querySelector(".dropdownContent")).toBeInTheDocument();
    fireEvent.click(dropdownIcon);
  });

  test("DropdownItem 클릭시 함수 동작", () => {
    const { getByText, container } = component;
    const dropdownIcon = container.querySelector(".dropdownIcon")!;

    expect(container.querySelector(".dropdownContent")).not.toBeInTheDocument();
    fireEvent.click(dropdownIcon);

    const firstItem = getByText("첫 번째 동작");
    fireEvent.click(firstItem);
    expect(playGround.innerText).toBe("첫 번째 버튼이 클릭되었습니다");
    expect(click1st).toHaveBeenCalledTimes(1);

    const secondItem = getByText("두 번째 동작");
    fireEvent.click(secondItem);
    expect(playGround.innerText).toBe("두 번째 버튼이 클릭되었습니다");
    expect(click1st).toHaveBeenCalledTimes(1);
  });
});
