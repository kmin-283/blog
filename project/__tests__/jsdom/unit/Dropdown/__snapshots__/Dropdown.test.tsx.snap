// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InpageNavigation 컴포넌트 렌더링 1`] = `
<section
  className="dropdown"
>
  <svg
    className="dropdownIcon"
    fill="currentColor"
    height="2em"
    onClick={[Function]}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 16 16"
    width="2em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M3 9.5a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm5 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3zm5 0a1.5 1.5 0 1 1 0-3 1.5 1.5 0 0 1 0 3z"
    />
  </svg>
</section>
`;
